CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java

```diff
@@ -70,38 +79,44 @@
 })
 public final class QualityControlViewPane extends BorderPane {
 
-    private final QualityControlViewTopComponent parent;
-    private final TableView<QualityControlEvent> qualityTable;
+    private static final Map<QualityControlRule, QualityCategory> rulePriorities = new HashMap<>();
+    private static final List<ToggleGroup> toggleGroups = new ArrayList<>();
     private final TableColumn<QualityControlEvent, QualityControlEvent> identifierColumn;
     private final TableColumn<QualityControlEvent, QualityControlEvent> typeColumn;
     private final TableColumn<QualityControlEvent, QualityControlEvent> qualityColumn;
     private final TableColumn<QualityControlEvent, QualityControlEvent> reasonColumn;
+    private final TableView<QualityControlEvent> qualityTable;
     private final FlowPane optionsPane;
 
-    public QualityControlViewPane(QualityControlViewTopComponent topComponent) {
-        this.parent = topComponent;
+    static {
+        for (final QualityControlRule rule : Lookup.getDefault().lookupAll(QualityControlRule.class)) {
+            rulePriorities.put(rule, rule.getCategory(0));
+        }
+    }
```

Comment:
Running this in a static block like this means this bit of code must run when the class is loaded into memory (i.e when Constellation loads). This becomes problematic for a couple of reasons:
1) Constellation load times will take longer.
2) There is actually a runtime exception that Java throws when there are too many things run in a static block like this and a while back I moved them to be called lazily instead which you can do in `getPriorities()` with something like:

```java
private static Map<QualityControlRule, QualityCategory> rulePriorities = null;
...
...
public static Map<QualityControlRule, QualityCategory> getPriorities() {
    if(rulePriorities == null) {
        for (final QualityControlRule rule : Lookup.getDefault().lookupAll(QualityControlRule.class)) {
            rulePriorities.put(rule, rule.getCategory(0));
        }
    }
}

return rulePriorities;
```

FYI the call to `Lookup` isn't super efficient. For instance I have been able to dramatically shave seconds off the execution just by caching the object returned by `Lookup` where as it was being called many times (like in for loops in other parts of the app).

Reply:
Hi @arcturus2 , Thanks for all the feedback! 👍 

>With the custom priorities radio option section, I don't believe this is being persisted into a Preferences object so it looks like changes will be lost after restarting Constellation.

You are correct, it doesn't currently.
I've been looking to store it into a preferences object but have hit a snag. I want to store a map into the preferences but have not found a method to set an object into the preferences. it only takes String or numeric values. 
From this I've pivoted to converting the map to a json String. I've looked through the utils and haven't found anything out the box. Do you have any suggestions re this or another possible solution?

>It didn't look like any of the "Rule" classes like UnknownTypeRule needed any changes; that's interesting and good actually.

Ideally this was what I was going for so I'm glad it all works without needing further extensions to those classes.


>Can you please write some unit tests to cover the logic? If you want a hand just let me know and happy to help writing uni tests.

>There should be a mention about this change to the users via the whatsnew.txt file informing them that scores are replaced by categories.

I can do these. 🍰 


>Running this in a static block like this means this bit of code must run when the class is loaded into memory (i.e when Constellation loads). 

Ohh no worries, I think that should be an easy fix.


>FYI the call to Lookup isn't super efficient. For instance I have been able to dramatically shave seconds off the execution just by caching the object returned by Lookup

I could store it into a class level variable, but when would i know to update it? only on instantiation of the view? 

I'll be a little while getting these changes in, and will re-request review once it's all wrapped up. 
Once again, cheers for taking such an in-depth look at this. 🙏

Reply:
>...From this I've pivoted to converting the map to a json String. I've looked through the utils and haven't found anything out the box. Do you have any suggestions re this or another possible solution?

Yep that is how some other preferences work when you need to do more than a String. Have a look at your boostrap folder and then `build\testuserdir\config\Preferences\au\gov\asd\tac\constellation\plugins.properties`. Its a json string which `RecentParameterValues` loads and saves.

> I could store it into a class level variable, but when would i know to update it? only on instantiation of the view?
Yep that's how I would do it. You should be fine to put it in the class where `getPriorities` gets called.